### ISSUE TITLE
detect.py Inference not working on GPU (uses CPU by default)

### ISSUE THREAD

**Issue by davendramaharaj1** [ultralytics/yolov5#11229]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and found no similar bug report.

### YOLOv5 Component

Detection

### Bug

I trained yolov5 on a custom dataset and I have the best.pt. I am using this weight file along with a test image and video to run inference. However, I am not sure if inference is being run on the cpu or my gpu since it takes a longer time than expected. I added some print statements to find out which device is used, which prints to the console "cuda:0". Here is an example of the results I received for a test image of size 416x416

`Saving ../results/test2/test/stage0_Conv_features.png... (32/48)
Saving ../results/test2/test/stage1_Conv_features.png... (32/96)
Saving ../results/test2/test/stage2_C3_features.png... (32/96)
Saving ../results/test2/test/stage3_Conv_features.png... (32/192)
Saving ../results/test2/test/stage4_C3_features.png... (32/192)
Saving ../results/test2/test/stage5_Conv_features.png... (32/384)
Saving ../results/test2/test/stage6_C3_features.png... (32/384)
Saving ../results/test2/test/stage7_Conv_features.png... (32/768)
Saving ../results/test2/test/stage8_C3_features.png... (32/768)
Saving ../results/test2/test/stage9_SPPF_features.png... (32/768)
Saving ../results/test2/test/stage10_Conv_features.png... (32/384)
Saving ../results/test2/test/stage11_Upsample_features.png... (32/384)
Saving ../results/test2/test/stage12_Concat_features.png... (32/768)
Saving ../results/test2/test/stage13_C3_features.png... (32/384)
Saving ../results/test2/test/stage14_Conv_features.png... (32/192)
Saving ../results/test2/test/stage15_Upsample_features.png... (32/192)
Saving ../results/test2/test/stage16_Concat_features.png... (32/384)
Saving ../results/test2/test/stage17_C3_features.png... (32/192)
Saving ../results/test2/test/stage18_Conv_features.png... (32/192)
Saving ../results/test2/test/stage19_Concat_features.png... (32/384)
Saving ../results/test2/test/stage20_C3_features.png... (32/384)
Saving ../results/test2/test/stage21_Conv_features.png... (32/384)
Saving ../results/test2/test/stage22_Concat_features.png... (32/768)
Saving ../results/test2/test/stage23_C3_features.png... (32/768)
image 1/1 /home/ubuntu/test.jpg: 416x416 1 bus, 2 persons, 18733.2ms
Speed: 0.2ms pre-process, 18733.2ms inference, 1.0ms NMS per image at shape (1, 3, 416, 416)
Results saved to ../results/test2
1 labels saved to ../results/test2/labels
`

### Environment

- YOLO: Yolov5 (master branch)
- OS: Ubuntu 20.04.5 LTS
- Python 3.8.10
- GPU: NVIDIA 06:00.0 NVIDIA Corporation Device 2236
- CPU: Intel Xeon Platinum 8358 (30) @ 2.593GHz

### Minimal Reproducible Example

python3 detect.py --weights ../best.pt --source ../test.jpg --img 416 --device 0 --visualize --save-txt --save-conf --save-crop --project ../results/ --name test

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@davendramaharaj1 hello! Thanks for reaching out. It looks like your system is recognizing the GPU (`cuda:0`), but the inference time is unusually long. Here are a few steps to troubleshoot:

1. **Check GPU Utilization**: Ensure that your GPU is being utilized during inference by monitoring it with tools like `nvidia-smi`.

2. **CUDA & cuDNN**: Verify that you have the correct versions of CUDA and cuDNN installed for your PyTorch version.

3. **PyTorch & Torchvision**: Make sure you're using compatible versions of PyTorch and torchvision with your CUDA version.

4. **Batch Size**: If you're running a single image, the GPU might be underutilized. Try increasing the batch size if applicable.

5. **Visualize Flag**: The `--visualize` flag will save model feature maps, which can be time-consuming. Try running without this flag to see if there's a significant speed difference.

6. **Precompiled Operations**: Some operations might not be optimized for your GPU. You can try to recompile these operations specifically for your GPU architecture.

7. **Update YOLOv5**: Ensure you're using the latest version of YOLOv5, as performance improvements are regularly added.

If the issue persists, please refer to our documentation for more detailed troubleshooting steps or consider opening a new issue with updated details. 😊
